rust/src/services/mod.rs

```diff
@@ -0,0 +1,179 @@
+        self.fetcher.model().await
+    }
+
+    pub async fn seed_dict(&self) -> Result<Option<Arc<SeedDict>>, FetchError> {
+        self.fetcher.seed_dict().await
+    }
+
+    pub async fn sum_dict(&self) -> Result<Option<Arc<SumDict>>, FetchError> {
+        self.fetcher.sum_dict().await
+    }
+
+    pub async fn handle_message(&self, message: Vec<u8>) -> Result<(), PetMessageError> {
+        self.message_handler.handle_message(message).await
+    }
+}
+
+impl PetService<Box<dyn Fetcher + Sync>, Box<dyn PetMessageHandler + Sync>> {
+    pub fn new(
+        event_subscriber: &EventSubscriber,
+        requests_tx: RequestSender<Traced<Request>>,
+    ) -> Self {
+        Self {
+            fetcher: fetcher(event_subscriber),
+            message_handler: message_handler(event_subscriber, requests_tx),
+        }
+    }
+}
+
+fn fetcher(event_subscriber: &EventSubscriber) -> Box<dyn Fetcher + Sync> {
+    let round_params = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(RoundParamsService::new(event_subscriber));
+
+    let mask_length = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(MaskLengthService::new(event_subscriber));
+
+    let scalar = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(ScalarService::new(event_subscriber));
+
+    let model = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(ModelService::new(event_subscriber));
+
+    let sum_dict = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(SumDictService::new(event_subscriber));
+
+    let seed_dict = ServiceBuilder::new()
+        .buffer(100)
+        .concurrency_limit(100)
+        .service(SeedDictService::new(event_subscriber));
+
+    Box::new(FetcherService::new(
+        round_params,
+        sum_dict,
+        seed_dict,
+        mask_length,
+        scalar,
+        model,
+    ))
+}
+
+fn message_handler(
+    event_subscriber: &EventSubscriber,
+    requests_tx: RequestSender<Traced<Request>>,
+) -> Box<dyn PetMessageHandler + Sync> {
+    // TODO: make this configurable. Users should be able to
+    // choose how many threads they want etc.
+    //
+    // TODO: don't unwrap
+    let thread_pool = Arc::new(ThreadPoolBuilder::new().build().unwrap());
+
+    let message_parser = ServiceBuilder::new()
+        // allow processing 10 request concurrently, and allow up
+        // to 10 requests to be pending. Once 10 requests are
+        // being processed and 10 are queued, the service will
+        // report that it's not ready.
+        //
+        // FIXME: what's a good concurrency limit? Should we limit
+        // the number of concurrent messages being processed to
+        // the number of threads in the rayon thread-pool? Or is
+        // rayon smart enough to avoid too many context switches?
+        .buffer(100)
```

Comment:
`buffer(10)` means that the requests go through a queue with 10 slots before reaching our service. A request is dequeued and handed to the service when the service is ready to process a new request.

`concurrency_limit(10)` limits the number of requests that our service can process concurrently to 10. This means that if the _inner_ service is already handling 10 requests (ie there are 10 Futures returned by `Service::call` currently running), the service will wait for a future to finish before accepting a new request.

The two are a bit redundant, but slightly different, and the order in which they wrap the service matter. See: https://docs.rs/tower/0.3.1/tower/builder/struct.ServiceBuilder.html#order

Reply:
thanks for the explanation and link. this means that the documentation is off and should read 100 instead of 10 because of the chosen service parameters.

Reply:
fixed.